### ISSUE TITLE
Connection pool doesn't like blocking commands?

### ISSUE THREAD

**Issue by cactorium** [informatikr/hedis#23]
```
{-# LANGUAGE OverloadedStrings #-}
import Control.Concurrent
import qualified Data.ByteString.Char8 as B
import Database.Redis

main :: IO ()
main = do
    conn <- connect defaultConnectInfo { connectMaxConnections = 1024 }
    forkIO $ do 
        threadDelay 100000
        runRedis conn $ lpush "locktest" ["0"]
        return ()
    bar <- runRedis conn $ blpop ["locktest"] 0
    case bar of
        Left foobar -> putStrLn "Test!"
        Right barfoo -> putStrLn "Potato!"
```

The program hangs, although it doesn't for smaller values in threadDelay. I was assuming the behavior would be for it to pause 0.1 seconds for the LPUSH and then print out "Potato!" after the BLPUSH unblocks. Is this intended behavior?

**Comment by vipo** [comment]
The problem is related to resource-pool stripes which is one by default. I have increased 1 to a magic number 20 and now blocking operations work well. 

This project needs some love :/

**Comment by ozataman** [comment]
We are experiencing an intermittent blocking issue while using hedis' pubsub implementation. The PubSub somehow stops working after long periods of idling. Somehow the receiving end no longer responds to commands sent via the channel.

I'm still not sure what the cause is, but this made me wonder if the stripe count is the culprit.

Have you had any more thought on this issue?

Also one question: Were you testing with +RTS -N4 or something to that effect? You probably need the threaded runtime to test this kind of thing.

Thank you.

**Comment by MichaelXavier** [comment]
Just tested out the original code with +RTS -N4 and it does indeed hang. The order of events is:

1. blpop checks out a connection and issues its ill-fated `blpop` request.
2. forked thread wakes up from its delay and issues its `lpush` request.

Its now deadlocked. It isn't clear if the lpush never gets to create a connection and stalls out there or if it stalls out in redis code. As a little test to check my understanding of pool, I created a contrived test with a fake resource where no connections are being made:

```haskell
module Main
    ( main
    ) where


import Data.Pool
import Control.Concurrent


type Connection = ()

connect :: IO (Pool Connection)
connect = createPool create destroy 1 maxIdle maxConns
  where maxIdle = 30
        maxConns = 1024
        create = putStrLn "create" >> return ()
        destroy _ = putStrLn "destroy"


-------------------------------------------------------------------------------
main :: IO ()
main = do
    pool <- connect
    forkIO $ do
        putStrLn "big sleep"
        threadDelay 100000
        putStrLn "trying to push"
        withResource pool $ \_ -> putStrLn "push!"
        putStrLn "pushed"
        return ()
    putStrLn "blpopping"
    withResource pool $ \_ -> threadDelay 5000000
```

The output is

```
blpopping
big sleep
create
trying to push
create
push!
pushed
```

As you can see, even though a thread is waiting with a connection checked out, a new connection gets created and the fake lpush goes through.

Its as though when a redis connection is being waited on, the runtime will not permit a thread context switch that needs to use the socket. Maybe pipelining is the issue here?

**Comment by MichaelXavier** [comment]
I think I've figured it out. I think it is pipelining, If we change this line:

```haskell
bar <- runRedis conn $ blpop ["locktest"] 0
```
to
```haskell
    bar <- runRedis conn $ do
      res <- blpop ["locktest"] 0
      liftIO (evaluate res)
```

The code behaves as expected: the blpop yields to the lpush, then immediately picks up its result and returns it. The docs say pipelining can even occur between separate `runRedis` calls (indeed even across threads) as long as the result isn't evaluated *within* the `runRedis` calls. Without the evaluation, I think the server is getting:

```
BLPOP locktest 0
LPUSH lpush
```
and thus never gets back the first response. However this would also be bad even without the contrived deadlock: you lazily blpop in one thread and do a completely unrelated op, say a `GET` or `SET` in another thread. The pipelining would put your other op behind `blpop` which for all we know will never get called again. Even if you specify say a 5s timeout for `blpop`, automatic pipelining would cause your other requests to wait up to 5s for blpop, thus negating any performance gained from pipelining to save on round trips.

I'm tempted to say that *any* of the explicitly blocking redis calls are not safe for pipelining, because unless their condition is met immediately they will hold up other commands. I bet you could even implement this by tacking `seq` into those commands to have them stop the currently active pipeline.

**Comment by ozataman** [comment]
Isn't pipelining per connection? Blpop checks out and blocks the connection from pool (via runRedis) so it should be impossible for any subsequent request from another runRedis to happen from the same connection anyway until blpop actually returns a result, right?

I may be missing something though... Like whether pipelining can span across multiple connection resources.

**Comment by MichaelXavier** [comment]
It is per connection. The docs say "Automatic pipelining also works across several calls to runRedis, as long as replies are only evaluated outside the runRedis block." I'm actually not sure how that mechanism works since runRedis calls `withResource` internally which supposedly checks out a new connection resource, but the fact that evaluating inside of the `runRedis` block stops the pipelining seems to support it.

**Comment by qrilka** [comment]
@ozataman regarding you notice on hanging pubsub - that looks like the situation described in #39 - could you probably comment there?

**Comment by ozataman** [comment]
This is unfortunately still happening. I think we were able to produce some
weird degenerate behavior, possibly due to the way pipelining is lazyly
implemented. Wasnt related to the resource pool issue.

@michaelxavier - do you remember the specifics?

**Comment by qrilka** [comment]
@ozataman if it was about `PubSub` then it wasn't fixed yet, I'm trying to understand if there could be any possible reason other than half-open sockets and about this issue talking about leaking connection which could lead to the demonstrated deadlock - I'm going to prepare new PR about it soon

**Comment by MichaelXavier** [comment]
@ozataman Unfortunately not. I don't think I specifically looked into pubsub, more into how automatic pipelining worked with standard operations and blocking operations.

**Comment by k-bx** [comment]
Fixed by PR #45
